### PR TITLE
Inline functions standing out in pprof

### PR DESCRIPTION
**Pull request by antiguru**
Inline timestamp implementations, specifically those that stand out in CPU profiles in incident 75: https://pprof.me/3db718b/ shows that `<mz_timely_util::order::Interval<> as timely::order::PartialOrder>::less_equal` has a separate stack frame, which means it's not inlined. This makes sense because we're implementing a trait from a different crate, and Rust's inlining often cannot reliably cross crate boundaries.

  * This PR adds a feature that has not yet been specified.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by umanwizard** [comment]
> has a separate stack frame, which means it's not inlined

That is not necessarily true. Rust debug symbols are able to represent inlined frames, and AFAIK PolarSignals renders them.

**Review by umanwizard — APPROVED** [review]
Anyway, it can't hurt to add these.

**Comment by petrosagg** [comment]
@antiguru I'm not sure these annotations do anything because the struct is generic and therefore the functions are only compiled when the crate using them is compiled too (so that the types are known) and therefore the optimizer will consider them for inlining anyway because they will be in the same compilation unit. Have you checked the produced assembly to see if anything changes?

**Comment by petrosagg** [comment]
> Anyway, it can't hurt to add these.

If it is true that the function was **not** inlined due to the compiler's decision then we should prove to ourselves that overriding the compiler's decision will help with a benchmark. Afaik there are costs that come with inlining (e.g instruction cache hit rate)

**Comment by umanwizard** [comment]
@antiguru , FYI, a feature has been added in PS that shows whether frames were inlined when you hover over the frame.

Indeed, `less_equal` is not inlined.

**Comment by antiguru** [comment]
I find Rust's inlining semantics hard to follow, but here's an attempt to explain what's happening:
* `PartialOrder` is a non-generic trait, which means for each type there is one implementation and it doesn't depend on the caller.
* `PartialOrder` is defined in the `timely` crate, and implemented in various places. `MzOffset` is in `mz-storage-types`, `Partitioned` sits in `mz-timely-util`.
* `reclock_time` sits in `mz-storage`.
* Rust can only inline within the same create, but not from a downstream crate to a previously defined symbol. This is the part I'm not confident about -- callgrind would tell us!

This would mean that we indeed need to inline the `PartialOrder` implementations.

**Comment by antiguru** [comment]
I updated the PR to inline the `PartialOrder` implementation for both `MzOffset` and `Partitioned`, because I don't know which we're using in this situation.

**Comment by guswynn** [comment]
curious, why `#[inline]` over `#[inline(always)]`?

**Comment by antiguru** [comment]
> curious, why `#[inline]` over `#[inline(always)]`?

It allows the compiler to decide! In some situations it can be better not to inline, and we're not precluding this option.

**Comment by guswynn** [comment]
Apparently `inline(always)` also lets the compiler decide, but less aggressively? unfortunately, what this means appears to be undocumented :(
